### PR TITLE
feat(p2p): avoid blocking ping/pong; feed msgs through a limited buffer to a separate goroutine

### PR DESCRIPTION
**Pull request by mhofman**
Refs: https://github.com/Agoric/agoric-sdk/issues/10742

## Description

Add a new goroutine to handle messages read from the peer connection. This allows processing ping/pong messages even if reactors are blocked. To avoid unbounded growth the buffered channel between the recv and msg processing go routines is limited in depth of number of msgs, and in aggregate size of msgs.

This maintains the in order processing of messages over a p2p connection, but offloads all its processing, including protobuf parsing to the goroutine, and as such localizes the changed needed to the connection abstraction unlike other cometbft efforts (https://github.com/cometbft/cometbft/pull/3230, https://github.com/cometbft/cometbft/pull/3209). It's also compatible with efforts to parallelize some reactor processing that are ongoing (https://github.com/cometbft/cometbft/issues/2685, https://github.com/cometbft/cometbft/pull/3554)

## Scaling considerations 

Incorporates a "fix" for https://github.com/cometbft/cometbft/issues/3237 as we need to be able to keep receiving messages on a channel that has a message pending in the queue. This is done by simply reallocating a new buffer instead of truncating the working one. I believe this was a premature optimization of the tendermint code, golang has a pretty good memory allocator.

The aggregate size of the queue is bounded to the default max size of a message in a channel, but allowing a single message over that default size if the channel needs it. I believe this ensure a remote peer cannot consume significantly more memory than it could otherwise: it can only at most consume twice the max size of a message allowed by any channel.

## Testing

Tested manually on a follower running this branch as an out-of-process tendermint.

TODO: add unit tests for edge conditions
- Test with `MsgRecvQueueSize` at 0 (unbuffered channel). A single message received should be processed by the reactor immediately
- Test with 2 messages in combination putting the queue over the aggregate size limit: verify that the second message blocks until the first message is processed by the consuming routine
- Test that a single message of the aggregate size limit or larger can be processed
- Test that closing the connection with messages still pending on the queue doesn't log any panic
- Test that closing the connection with data remaining to be read from the raw connection doesn't log any panic (should be an existing test?)

---

#### PR checklist

- [ ] Tests written/updated
- [ ] Changelog entry added in `.changelog` (we use
  [unclog](https://github.com/informalsystems/unclog) to manage our changelog)
- [ ] Updated relevant documentation (`docs/` or `spec/`) and code comments

### CODE REVIEW

File: p2p/conn/connection.go [hunk 1]
```diff
@@ -664,6 +692,33 @@ FOR_LOOP:
...
+			break
+		}
+
+		msgLen := len(connMsg.msgBytes)
+		newBufferSize := c.msgRecvSize.Add(^uint32(msgLen - 1)) // Sub(msgLen)
+		c.Logger.Debug("Processing message", "chID", connMsg.channelID, "msgBytesBufferedLength", newBufferSize, "pendingMsgs", len(c.msgRecvQueue))
+
+		// Wake up recvRoutine if necessary
+		select {
+		case c.msgRecvAck <- struct{}{}:
```
Comment (Muneeb147) [inline p2p/conn/connection.go]: Couldn't it be just a bool channel instead? OR are we using struct{}{} as it is memory efficient (0 bytes) when it comes to just signalling things with channel?

Comment (JeancarloBarrios) [reply p2p/conn/connection.go]: Go optimizes an empty struct not to use memory, so it is technically more memory efficient and idiomatic. In Go you use an empty structure when you want to use a channel for signaling only.

File: p2p/conn/connection.go [hunk 2]
```diff
@@ -645,9 +662,20 @@ FOR_LOOP:
...
 			}
 			if msgBytes != nil {
-				c.Logger.Debug("Received bytes", "chID", channelID, "msgBytes", msgBytes)
-				// NOTE: This means the reactor.Receive runs in the same thread as the p2p recv routine
-				c.onReceive(channelID, msgBytes)
+				msgLen := uint32(len(msgBytes))
+				bufferMaxSize := uint32(defaultRecvMessageCapacity)
+				if msgLen > bufferMaxSize {
+					bufferMaxSize = msgLen
+				}
```
Comment (Muneeb147) [inline p2p/conn/connection.go:683]: @mhofman
Will this ever be True at this place? 🤔 

The function `recvPacketMsg` at line no. 656 must already throw err (`received message exceeds available capacity`) if `msgLen > bufferMaxSize` was true.

So I think this condition is redundant and will always evaluate to `false` OR I might be missing something.

File: p2p/conn/connection.go [hunk 3]
```diff
@@ -45,13 +45,19 @@ const (
 	defaultSendTimeout         = 10 * time.Second
 	defaultPingInterval        = 60 * time.Second
 	defaultPongTimeout         = 45 * time.Second
+	defaultMsgRecvQueueSize    = 1000
```
Comment (Muneeb147) [inline p2p/conn/connection.go:48]: Have we picked this size based on some testing/analysis?

File: p2p/conn/connection.go (hunk 2 above)
Comment (mhofman) [reply p2p/conn/connection.go:683]: My understanding is that a channel can be configured with a different message size limit, which can be higher than the default. If such a message were to be received on the channel, the connection needs to allow it.

File: p2p/conn/connection.go (hunk 3 above)
Comment (mhofman) [reply p2p/conn/connection.go:48]: This was a guess. I have been running a mainnet follower since with a higher limit and will be looking at the high water mark to get an idea of a reasonable default value.

File: p2p/conn/connection.go (hunk 2 above)
Comment (Muneeb147) [reply p2p/conn/connection.go:683]: Aah I see, `FillDefaults` conditionally fills Channel's `RecvMessageCapacity` with `defaultRecvMessageCapacity`. Got it!

Btw shouldn't we respect (or use) `channel.desc.RecvMessageCapacity` as our `bufferMaxSize` especially when it is set lesser than the default (`defaultRecvMessageCapacity`)?

Comment (Muneeb147) [reply p2p/conn/connection.go:683]: I believe defaultRecvMessageCapacity is rightly used as our goal is to keep total msgBytes in queue (at any instance) to be within 21Mb, right? Otherwise we'd be burning 1000*21 memory.

So I guess we can resolve the discussion. Things are clarified 👍

File: p2p/conn/connection.go [hunk 4]
```diff
@@ -136,6 +145,11 @@ type MConnConfig struct {
 
 	// Maximum wait time for pongs
 	PongTimeout time.Duration `mapstructure:"pong_timeout"`
+
+	// Maximum number of received messages to buffer from a peer
+	MsgRecvQueueSize int `mapstructure:"msg_recv_queue_size"`
+
+	RecvMessageCapacity int `mapstructure:"recv_message_capacity"`
```
Comment (gibson042) [inline p2p/conn/connection.go:153]: Please add a docstring here too; the units of `RecvMessageCapacity` are not locally clear.

File: p2p/conn/connection.go [hunk 5]
```diff
@@ -645,9 +675,19 @@ FOR_LOOP:
...
+				bufferMaxSize := uint32(c.config.RecvMessageCapacity)
+				if msgLen > bufferMaxSize {
+					bufferMaxSize = msgLen
+				}
+				c.msgRecvSize += msgLen
+				c.Logger.Debug("Received bytes", "chID", channelID, "msgBytesLength", msgLen, "msgBytesBufferedLength", c.msgRecvSize, "pendingMsgsAhead", len(c.msgRecvQueue))
+				// Wait until our message fits in the queue
+				for c.msgRecvSize > bufferMaxSize {
+					c.msgRecvSize -= <-c.msgRecvAck
+				}
```
Comment (gibson042) [inline p2p/conn/connection.go]: `msgRecvSize` seems misnamed, something like `msgRecvBufferedByteCount` or `msgBytesBufferedLength` (matching the log message) would be more better IMO.

File: p2p/conn/connection.go [hunk 6]
```diff
@@ -664,6 +704,31 @@ FOR_LOOP:
...
 }
 
+func (c *MConnection) recvMsgRoutine() {
+	defer c._recover()
+
+	for {
+		connMsg, ok := <-c.msgRecvQueue
+
+		if !ok {
+			break
```
Comment (gibson042) [inline p2p/conn/connection.go:716]: When is this `break` expected to be encountered? If it's an error condition, something noisier is warranted. But if it's just routine shutdown from `recvRoutine` closing the channel, a comment to that effect would be nice.

File: p2p/conn/connection.go [hunk 7]
```diff
@@ -664,6 +704,31 @@ FOR_LOOP:
...
+		}
+
+		msgLen := len(connMsg.msgBytes)
+		// Will never block as long as msgRecvAck is at least one bigger than msgRecvQueue
+		c.msgRecvAck <- uint32(msgLen)
```
Comment (gibson042) [inline p2p/conn/connection.go:721]: :+1:

File: p2p/conn/connection.go [hunk 8]
```diff
@@ -664,6 +704,31 @@ FOR_LOOP:
...
+		// The following logging of c.msgRecvSize is not a correct representation of the current aggregate size
+		// of the queue but is instead the size of the queue as accounted by the producer routine (trailing behind)
```
Comment (gibson042) [inline p2p/conn/connection.go]: "producer routine"?

File: p2p/conn/connection.go [hunk 9]
```diff
@@ -664,6 +704,31 @@ FOR_LOOP:
...
+		// However for debug logging it may be enough
+		c.Logger.Debug("Processing message", "chID", connMsg.channelID, "msgBytesBufferedLength", c.msgRecvSize, "pendingMsgs", len(c.msgRecvQueue))
+
+		// NOTE: This means the reactor.Receive of all channels run in the same thread
+		c.onReceive(connMsg.channelID, connMsg.msgBytes)
+	}
+
+	close(c.msgRecvAck)
```
Comment (gibson042) [inline p2p/conn/connection.go]: Should this instead be moved up to the top of function?
```go
func (c *MConnection) recvMsgRoutine() {
	defer c._recover()
	defer close(c.msgRecvAck)

	…
}
```

File: p2p/conn/connection.go [hunk 10]
```diff
@@ -867,11 +932,7 @@ func (ch *Channel) recvPacketMsg(packet tmp2p.PacketMsg) ([]byte, error) {
 	if packet.EOF {
 		msgBytes := ch.recving
 
-		// clear the slice without re-allocating.
-		// http://stackoverflow.com/questions/16971741/how-do-you-clear-a-slice-in-go
-		//   suggests this could be a memory leak, but we might as well keep the memory for the channel until it closes,
-		//	at which point the recving slice stops being used and should be garbage collected
-		ch.recving = ch.recving[:0] // make([]byte, 0, ch.desc.RecvBufferCapacity)
+		ch.recving = make([]byte, 0, ch.desc.RecvBufferCapacity)
```
Comment (gibson042) [inline p2p/conn/connection.go:932]: :+1:

File: p2p/conn/connection_test.go [hunk 11]
```diff
@@ -43,6 +43,19 @@ func createMConnectionWithCallbacks(
...
+	onError func(r interface{}),
+	cfg MConnConfig,
+
+) *MConnection {
+	chDescs := []*ChannelDescriptor{{ID: 0x01, Priority: 1, SendQueueCapacity: 1}}
+	c := NewMConnectionWithConfig(conn, chDescs, onReceive, onError, cfg)
+	c.SetLogger(log.TestingLogger())
+	return c
+}
+
```
Comment (gibson042) [inline p2p/conn/connection_test.go]: `createMConnectionWithConfig` seems to be unused.
```suggestion
```

**Review by gibson042 — APPROVED** [review]

File: p2p/conn/connection.go (hunk 6 above)
Comment (mhofman) [reply p2p/conn/connection.go:716]: routine shutdown

File: p2p/conn/connection.go (hunk 8 above)
Comment (mhofman) [reply p2p/conn/connection.go]: The producer for this channel is `recvRoutine`. I admit we could be better at names here

Comment (JeancarloBarrios) [reply p2p/conn/connection.go]: Should we update the names?

Comment (mhofman) [reply p2p/conn/connection.go]: I wanted to avoid changing the name of the original routine to keep future conflict resolution to a minimum, but maybe that's a misplaced hope

Comment (JeancarloBarrios) [reply p2p/conn/connection.go]: ok !!

**Review by JeancarloBarrios — APPROVED** [review]

File: p2p/conn/connection.go (hunk 3 above)
Comment (mhofman) [reply p2p/conn/connection.go:48]: @JeancarloBarrios please update this to 5000
